src/MagnumPlugins/AssimpImporter/AssimpImporter.cpp

```diff
@@ -1118,6 +1135,235 @@ Containers::Optional<ImageData2D> AssimpImporter::doImage2D(const UnsignedInt id
+       Blender doesn't really support cinematic animations (affecting multiple
+       objects): https://blender.stackexchange.com/q/5689. And since
+       https://github.com/KhronosGroup/glTF-Blender-Exporter/pull/166, these
+       are exported as a set of object-specific clips, which may not be wanted,
+       so we give the users an option to merge them all together. */
+    const std::size_t animationBegin =
+        configuration().value<bool>("mergeAnimationClips") ? 0 : id;
+    const std::size_t animationEnd =
+        configuration().value<bool>("mergeAnimationClips") ? _f->scene->mNumAnimations : id + 1;
+
+    /* Calculate total data size and track count. If merging all animations together,
+       this is the sum of all clip track counts. */
+    std::size_t dataSize = 0;
+    std::size_t trackCount = 0;
+    for (std::size_t a = animationBegin; a != animationEnd; ++a) {
+        const aiAnimation* animation = _f->scene->mAnimations[a];
+        for (std::size_t c = 0; c != animation->mNumChannels; ++c) {
+            const aiNodeAnim* channel = animation->mChannels[c];
+
+            /** @todo handle alignment once we do more than just four-byte types */
+            dataSize += channel->mNumPositionKeys*(sizeof(Float) + sizeof(Vector3)) +
+                channel->mNumRotationKeys*(sizeof(Float) + sizeof(Quaternion)) +
+                channel->mNumScalingKeys*(sizeof(Float) + sizeof(Vector3));
+
+            trackCount += (channel->mNumPositionKeys > 0 ? 1 : 0) +
+                (channel->mNumRotationKeys > 0 ? 1 : 0) +
+                (channel->mNumScalingKeys  > 0 ? 1 : 0);
+        }
+    }
+
+    /* Populate the data array */
+    Containers::Array<char> data{dataSize};
+
+    /* Import all tracks */
+    bool hadToRenormalize = false;
+    std::size_t dataOffset = 0;
+    std::size_t trackId = 0;
+    Containers::Array<Trade::AnimationTrackData> tracks{trackCount};
+    for(std::size_t a = animationBegin; a != animationEnd; ++a) {
+        const aiAnimation* animation = _f->scene->mAnimations[a];
+        for(std::size_t c = 0; c != animation->mNumChannels; ++c) {
+            const aiNodeAnim* channel = animation->mChannels[c];
+            const UnsignedInt target = doObject3DForName(channel->mNodeName.C_Str());
+            if(target == -1) {
+                Error{} << "Trade::AssimpImporter::animation(): unknown target node " << channel->mNodeName.C_Str();
+                return Containers::NullOpt;
+            }
+
+            /* Assimp only supports linear interpolation. For glTF splines
+               it simply uses the spline control points. */
+            /** @todo Before https://github.com/assimp/assimp/commit/e3083c21f0a7beae6c37a2265b7919a02cbf83c4
+             *        Assimp incorrectly read the spline tangents as values.
+             *        Is there a way to detect this so we can work around it,
+             *        or at least emit a warning?
+             */
```

Comment:
Sigh :) When all sane attempts failed, this is how I was detecting what library version I was linking to, by checking for a presence of an API that got added in given version: https://github.com/mosra/magnum-plugins/blob/ae1278de39a2b8b8f46dd590d39a15d0b410f18e/src/MagnumPlugins/AssimpImporter/Test/CMakeLists.txt#L46-L52

An even more insane variant of this is for SPIRV-Tools, which for some reason don't provide *any* version info whatsoever, so I have to detect each API addition with that: https://github.com/mosra/magnum-plugins/blob/ae1278de39a2b8b8f46dd590d39a15d0b410f18e/src/MagnumPlugins/SpirvToolsShaderConverter/CMakeLists.txt#L68-L88

- - -

Looking around, https://github.com/assimp/assimp/pull/3510 seems to be a PR that got merged around the same time as the above commit (few days earlier), so i think you could ignore the few commits in between and assume that if `aiScene` has the name member, it should also have non-fucked-up tangents? (I didn't really look further, maybe there's a better candidate tho.) The `CheckAssimpVersion5.cpp` file is currently in the test directory, I think it would make sense to move it one level up next to the plugin and add this other check to it similarly as with SPIRV-Tools.

Reply:
First things first: Thanks for the feedback so far. Always a pleasure getting great, detailed technical (and friendly!) feedback from you 🤗

On to the topic: The problem I see here is that even if we find out the Assimp version, I don't see a way (or maybe I missed it) to figure out if the animation data is taken from cubic splines in the first place. The only thing Assimp gives us is the keys and values, but no info about the interpolation function. A **terrible** workaround would be to manually parse the GLTF enough to find out if cubic spline interpolation is used, and then at least emit a warning.

Reply:
> I don't see a way to figure out if the animation data is taken from cubic splines in the first place.

Ahah, damn, I see. This sucks. So, if I understand correctly, on older versions this will give back a total nonsense, and on newer versions it'll ignore the tangent part of the cubic splines, right? I would definitely mention this in the limitations/caveats section of the docs.

I'd say in older versions you could grep the input file for a presence of `"CUBICSPLINE"` and then emit a warning in each `doAnimation()` if that's the case. I.e., doing something like this in `doOpenData()`:

```cpp
if(_f->importerIsGltf)
    _f->animationsProbablyBroken = Containers::StringView{data}.contains("\"CUBICSPLINE\""_s);
```

It's the laziest possible solution but should do the trick. It's probably also quite slow for large files, but the bar set by Assimp wasn't really high either. We're not here to fix the world, anyway.

In practice, I'm telling everyone to never use AssimpImporter for glTF because the material, light and mesh import sucks as well, so *optimistically* the only place this would affect is the tests, where you'll get various levels of bullshit from the TinyGltfImporter test files. And *hopefully* FBX animation support is not this bad, as that's what this importer is mainly for :)

Reply:
I went ahead and implemented the check, but then realized that `doOpenFile` fills the assimp scene data and then calls `doOpenData({})`, so for files this doesn't work. I don't feel like going full clown mode and trying to intercept this (custom IO callback? don't even know if this works) is worth it in the slightest, but I'm open to other suggestions.

Reply:
Sigh, again. Then ... I'd say print the warning on older versions always? It would mean some false positives, but better than being completely silent while giving out bad data.

There already is a callback implemented, for Magnum file callbacks, but I don't feel the extra complexity is worth it. Just warn always.